### PR TITLE
[WIP] Add WidgetSet endpoint

### PR DESCRIPTION
**Pull request by lpichler**
Fixes https://github.com/ManageIQ/manageiq-api/issues/953

This adds WidgetSet endpoint with CRUD.
#### LIST

```
GET /api/widget_sets
=>
{
    "name": "widget_sets",
    "count": 8,
    "subcount": 8,
    "pages": 1,
    "resources": [
        {
            "href": "http://localhost:3090/api/widget_sets/155"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/161"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/162"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/168"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/170"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/152"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/153"
        },
        {
            "href": "http://localhost:3090/api/widget_sets/154"
        }
    ],
    "actions": [
        {
            "name": "query",
            "method": "post",
            "href": "http://localhost:3090/api/widget_sets"
        },
        {
            "name": "create",
            "method": "post",
            "href": "http://localhost:3090/api/widget_sets"
        },
        {
            "name": "edit",
            "method": "post",
            "href": "http://localhost:3090/api/widget_sets"
        },
        {
            "name": "delete",
            "method": "post",
            "href": "http://localhost:3090/api/widget_sets"
        },
        {
            "name": "copy",
            "method": "post",
            "href": "http://localhost:3090/api/widget_sets"
        },
        {
            "name": "delete",
            "method": "delete",
            "href": "http://localhost:3090/api/widget_sets"
        }
    ],
    "links": {
        "self": "http://localhost:3090/api/widget_sets?offset=0",
        "first": "http://localhost:3090/api/widget_sets?offset=0",
        "last": "http://localhost:3090/api/widget_sets?offset=0"
    }
}
```

#### CREATE:
```
POST /api/widget_sets
{
    "name": "V",
    "description": "a",
    "set_data": { "col1": [7, 22, 20] },
    "group"   : {"id" : 1} # or href
}
=>
 {
    "results": [
        {
            "href": "http://localhost:3090/api/widget_sets/167",
            "id": "167",
            "name": "V",
            "description": "a",
            "set_type": "MiqWidgetSet",
            "created_on": "2020-12-09T15:29:35Z",
            "updated_on": "2020-12-09T15:29:35Z",
            "guid": "847f3d4f-9ccf-4049-9b60-d40a3d755bbd",
            "read_only": null,
            "set_data": {
                "col1": [
                    7,
                    22,
                    20
                ],
                "col2": [],
                "col3": [],
                "reset_upon_login": false,
                "locked": false
            },
            "mode": null,
            "owner_type": "MiqGroup",
            "owner_id": "1",
            "userid": null,
            "group_id": "1"
        }
    ]
}
```
#### EDIT:
```

POST /api/widget_sets/1
{
   "action" : "edit",
   "resource" : {
        "description": "testaa",
        "set_data": { "col1": [7, 22, 20] }
   } 
}

```
#### COPY
```
POST /api/widget_sets/167
{
   "action" : "copy",
   "resource" : {
        "name" : "ddd",
        "description": "testaa",
        "set_data": { "col1": [7, 22, 20] }
   } 
}
=>
{
    "success": true,
    "message": "Dashboard ddd successfully created.",
    "href": "http://localhost:3090/api/widget_sets/167"
}


```

#### DELETE

```
POST /api/widget_sets/167
{
   "action" : "delete"
}
=>

{
    "success": true,
    "message": "Dashboard Vd has been successfully deleted.",
    "href": "http://localhost:3090/api/widget_sets/169"
}

```

### Links
Requires:

- [x] https://github.com/ManageIQ/manageiq/pull/20890 - Validations
- [ ] https://github.com/ManageIQ/manageiq/pull/20972

### CODE REVIEW

**Comment by skateman** [comment]
@lpichler we also need an endpoint for copying I think

**Comment by lpichler** [comment]
> @lpichler we also need an endpoint for copying I think
> 
> https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/report_controller/dashboards.rb#L61

It could be done by with create action but I can add it to simplify copy action.

**Comment by skateman** [comment]
@lpichler as there's a separate method on the model for copying, I'd rather expose it than simulate it with a create...

**Comment by Fryguy** [comment]
I'm not sure I like all of this in the API...feels like a lot of the setup should be in the model as class methods. @agrare @abellotti Thoughts?  Is this a pattern we do elsewhere?

File: app/controllers/api/widget_sets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,134 @@
+module Api
+  class WidgetSetsController < BaseController
+    REQUIRED_FIELDS_TO_COPY = %w[name].freeze
+    ALLOWED_FIELDS = REQUIRED_FIELDS_TO_COPY + %w[group description guid read_only set_data].freeze
+    SET_DATA_COLS = %i[col1 col2 col3].freeze
+
+    def init_set_data(set_data)
```
Comment (Fryguy) [inline app/controllers/api/widget_sets_controller.rb:7]: Anything that's not exposed to a route should be moved into a private method.

File: app/controllers/api/widget_sets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    def init_widget_set(data)
+      data['set_type']   = "MiqWidgetSet"
+      data['owner_type'] = "MiqGroup"
+      data['set_data']   = init_set_data(data['set_data'])
+      data
+    end
+
+    def validate_description(group_id, description)
+      widget_set_exists_in_group = MiqWidgetSet.exists?(:owner_id => group_id, :description => description)
+      raise ArgumentError, "Description(Tab Title) must be unique for this group" if widget_set_exists_in_group
```
Comment (Fryguy) [inline app/controllers/api/widget_sets_controller.rb:27]: Like this stuff should be model validations...not in the API.

File: app/controllers/api/widget_sets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+      filtered_widget_ids = MiqWidget.where(:id => widget_ids).pluck(:id)
+      unless filtered_widget_ids.count == widget_ids.count
+        raise ArgumentError, "Unable to find widget set ids: #{widget_ids - filtered_widget_ids} "
+      end
+    end
+
+    def create_resource(type, id = nil, data = nil)
+      raise_if_unsupported_fields_passed(data)
+
+      raise ArgumentError, "Name cannot contain \"|\"" if data['name'].index('|')
```
Comment (Fryguy) [inline app/controllers/api/widget_sets_controller.rb:43]: Also should be a model validation.

**Comment by lpichler** [comment]
> I'm not sure I like all of this in the API...feels like a lot of the setup should be in the model as class methods. @agrare @abellotti Thoughts? Is this a pattern we do elsewhere?

Yes, this is exactly what I wanted to discuss -  I am not against, I wanted to see opinions from others. 
This is common that there is lot of validations of various kinds in controllers and when we are creating API endpoints we have to decide if validations should go to in API or to model level(and we can rely that data in db are already valid ).


cc @kbrock

File: app/controllers/api/widget_sets_controller.rb (hunk 3 above)
Comment (abellotti) [reply app/controllers/api/widget_sets_controller.rb:43]: Agreed, API does not raise any ArgumentErrors today, those come from the model.

File: app/controllers/api/widget_sets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    def init_set_data(set_data)
+      set_data ||= {}
+      new_set_data ||= {}
+      new_set_data[:col1] = set_data['col1'] || []
+      new_set_data[:col2] = set_data['col2'] || []
+      new_set_data[:col3] = set_data['col3'] || []
+      new_set_data[:reset_upon_login] = !!set_data['reset_upon_login']
+      new_set_data[:locked] = !!set_data['locked']
+      new_set_data
+    end
```
Comment (abellotti) [inline app/controllers/api/widget_sets_controller.rb:16]: I'm confused with this, is the API defining a structure beyond the basic MiqWidgetSet ?   It almost feels we need a new subclass of MiqWidgetSet in the model for what we're building/managing here and move most of this logic in there.

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-api/pull/968, https://github.com/ManageIQ/manageiq/pull/20890 including https://github.com/ManageIQ/manageiq-api/pull/968, https://github.com/ManageIQ/manageiq/pull/20890

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq, https://github.com/ManageIQ/manageiq-api including https://github.com/ManageIQ/manageiq-api/pull/968, https://github.com/ManageIQ/manageiq/pull/20890

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq, https://github.com/ManageIQ/manageiq-api including https://github.com/ManageIQ/manageiq-api/pull/968, https://github.com/ManageIQ/manageiq/pull/20890

File: app/controllers/api/widget_sets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      group_id = data['id']&.to_i || parse_resource_from(data['group']) || widget_set.group_id
+
+      widget_set = MiqWidgetSet.copy_dashboard(widget_set, data['name'], data['description'], group_id)
+
+      result = action_result(true, "Dashboard #{data['name']} successfully created.")
+      add_href_to_result(result, type, widget_set.id)
+      result
+    end
+
+    def parse_resource_from(attributes)
```
Comment (skateman) [inline app/controllers/api/widget_sets_controller.rb:47]: Shouldn't these two below be `private`?

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq, https://github.com/ManageIQ/manageiq-api including https://github.com/ManageIQ/manageiq-api/pull/968, https://github.com/ManageIQ/manageiq/pull/20890

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-api/pull/968  including https://github.com/ManageIQ/manageiq/pull/20890

**Comment by lpichler** [comment]
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq/pull/20890

File: app/controllers/api/widget_sets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+module Api
+  class WidgetSetsController < BaseController
+    REQUIRED_FIELDS_TO_COPY = %w[name].freeze
+    ALLOWED_FIELDS = REQUIRED_FIELDS_TO_COPY + %w[group description guid read_only set_data].freeze
```
Comment (abellotti) [inline app/controllers/api/widget_sets_controller.rb:4]: with 'read_only' here, does this mean the API can create read_only widget sets and cannot delete them ?  I usually think of read_only as set to true for some system seeded resources.

**Comment by abellotti** [comment]
needs a bulk query test. can be added in spec/requests/collections_spec.rb

**Comment by kavyanekkalapu** [comment]
@jrafanie This is needed for Overview/Reports/Dashboards -> Copy Selected Dashboard Missing endpoint for manipulating with dashboards form. Right now we are using miqajaxmethod as we don't have api.

**Pull request closed without merging**